procrustes/psdp.py

```diff
@@ -31,8 +31,164 @@
+    - Construct :math:`\hat{\mathbf{S}} = \mathbf{\Phi} \ast (
+        \mathbf{U}^T_1 \mathbf{F} \mathbf{V}_1 \mathbf{\Sigma}
+        + \mathbf{\Sigma} \mathbf{V}^T_1 \mathbf{F}^T \mathbf{U}_1)`.
+    - Perform spectral decomposition of :math:`\hat{\mathbf{S}}`.
+    - Computing intermediate matrices $\mathbf{P}_{11}$ and $\mathbf{P}_{12}$.
+    - Check if solution exists.
+    - Compute $\hat{\mathbf{P}}$ (required minimizer) using $\mathbf{P}_{11}$ and
+        $\mathbf{P}_{12}$.
+
+    References
+    ----------
+    .. [1] Jingjing Peng et al (2019). "Solution of symmetric positive semidefinite Procrustes
+        problem". Electronic Journal of Linear Algebra, ISSN 1081-3810. Volume 35, pp. 543-554.
+    """
+    # Check inputs and define the matrices F (matrix to be transformed) and
+    # G (the target matrix).
+    g, f = setup_input_arrays(
+        a,
+        b,
+        unpad_col,
+        unpad_row,
+        pad,
+        translate,
+        scale,
+        check_finite,
+        weight,
+    )
+    if g.shape != f.shape:
+        raise ValueError(
+            f"Shape of A and B does not match: {g.shape} != {f.shape} "
+            "Check pad, unpad_col, and unpad_row arguments."
+        )
+
+    # Perform Singular Value Decomposition (SVD) of G (here g).
+    u, singular_values, v_transpose = np.linalg.svd(g, full_matrices=True)
+    r = len(singular_values)
+    u1, u2 = u[:, :r], u[:, r:]
+    v1 = v_transpose.T[:, :r]
+    sigma = np.diag(singular_values)
+
+    # Representing the intermediate matrix S.
+    phi = np.array(
+        [[1 / (i**2 + j**2) for i in singular_values] for j in singular_values]
+    )
+    s = np.multiply(
+        phi, multi_dot([u1.T, f, v1, sigma]) + multi_dot([sigma, v1.T, f.T, u1])
+    )
+
+    # Perform Spectral Decomposition on S (here named s).
+    eigenvalues, unitary = np.linalg.eig(s)
+    eigenvalues_pos = [max(0, i) for i in eigenvalues]
+
+    # Computing intermediate matrices required to construct the required
+    # optimal transformation.
+    p11 = multi_dot([unitary, np.diag(eigenvalues_pos), np.linalg.inv(unitary)])
+    p12 = multi_dot([np.linalg.inv(sigma), v1.T, f.T, u2])
+
+    # Checking if solution is possible.
+    if np.linalg.matrix_rank(p11) != np.linalg.matrix_rank(
+        np.concatenate((p11, p12), axis=1)
+    ):
+        raise ValueError(
+            "Rank mismatch. Symmetric positive semidefinite Procrustes problem has no solution."
+        )
+
+    # Finding the required minimizer (optimal transformation).
+    mid1 = np.concatenate((p11, p12), axis=1)
+    mid2 = np.concatenate((p12.T, multi_dot([p12.T, np.linalg.pinv(p11), p12])), axis=1)
+    mid = np.concatenate((mid1, mid2), axis=0)
+    p = multi_dot([u, mid, u.T])
+
+    # Returning the result as a ProcrastesResult object.
+    return ProcrustesResult(
+        new_a=a,
+        new_b=b,
+        error=np.linalg.norm(f - np.dot(p, g)),
```

Comment:
We should also fix the error computing to use the `compute_error` function in https://github.com/banrovegrie/procrustes/blob/312da41d1cbe69c786db7e1f8516acfeed461bf1/procrustes/psdp.py#L202. This would help the maintenance in the future.

Reply:
@FanwangM there is an issue with using `compute_error` function. According to the papers (Woodgate, Peng), error is being calculated as `sqrt(compute_error)`.

So, in our implementation, should we use `compute_error` directly? Or should we use `sqrt` of it?

Reply:
For now, I am using the `sqrt` one to keep it consistent with the tests. Else we could always update the tests and remove `sqrt`.

Reply:
Within the object of `ProcrustesResult`, we should stick to the error itself, which is `compute_error`. But does it change the behavior of the Woodgate algorithm? I think it's OK for Peng's algorithm, @banrovegrie

Reply:
We should fix the tests instead.

Reply:
Ok, the behaviour of Woodgate doesn't change. It's just that we are calculating $\|. \|_F^2$ rather than $\|.\|_F$.

Reply:
Originally, I was thinking it may change the convergence rate or number of iterations, https://github.com/theochem/procrustes/blob/52d84d4c7920b08d3e47ecb0833fe8cc504f4cb6/procrustes/psdp.py#L205. Can you do a testing by counting the number of iterations locally?

Reply:
Oh yeah for the convergence rate i am using sqrt of `compute_error`. But while pushing the error as a part of `ProcrustesResult()` I am defaulting it to just `compute_error`.

Reply:
fixed in [171a794](https://github.com/theochem/procrustes/pull/180/commits/171a794f55ef10c0a2f5ad17240fdb094fd10bef)